### PR TITLE
Issue 2445: Expose ES rest client connect and socket timeout config options

### PR DESCRIPTION
**Pull request by twilmes**
Addresses: Issue 2445

Signed-off-by: Ted Wilmes <ted.wilmes@experoinc.com>

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### CODE REVIEW

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java [hunk 1]
```diff
@@ -267,6 +267,16 @@
                 "It is recommended to always enable index store names cache unless you have more then 50000 indexes " +
                 "per index store.", ConfigOption.Type.MASKABLE, true);
 
+    public static final ConfigOption<Integer> CONNECT_TIMEOUT =
+        new ConfigOption<>(ELASTICSEARCH_NS, "connect-timeout",
+            "Sets the maximum connection timeout (in milliseconds).", ConfigOption.Type.MASKABLE,
+            Integer.class, 1000);
```
Comment (li-boxuan) [inline janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java:273]: Why 1000? The default value for Apache HTTP client seems to be -1.

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java [hunk 2]
```diff
@@ -267,6 +267,16 @@
...
+
+    public static final ConfigOption<Integer> SOCKET_TIMEOUT =
+        new ConfigOption<>(ELASTICSEARCH_NS, "socket-timeout",
+            "Sets the maximum socket timeout (in milliseconds).", ConfigOption.Type.MASKABLE,
+            Integer.class, 30000);
```
Comment (li-boxuan) [inline janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java:278]: Same here. The default value for the Apache HTTP client seems to be -1.

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestClientSetup.java [hunk 3]
```diff
@@ -122,7 +123,26 @@ protected RestElasticSearchClient getElasticSearchClient(RestClient rc, int scro
...
+
+        final List<RequestConfigCallback> callbackList = new LinkedList<>();
+
+        final Integer connectTimeout = config.get(ElasticSearchIndex.CONNECT_TIMEOUT);
+        final Integer socketTimeout = config.get(ElasticSearchIndex.SOCKET_TIMEOUT);
+
+        callbackList.add((requestConfigBuilder) ->
+            requestConfigBuilder.setConnectTimeout(connectTimeout).setSocketTimeout(socketTimeout));
+
+        if (callbackList.isEmpty()) {
```
Comment (li-boxuan) [inline janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestClientSetup.java:135]: Seems this condition always evaluates to false.

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java (hunk 1 above)
Comment (twilmes) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java:273]: @li-boxuan thanks for the review. I'm not as familiar with the more recent ES updates but isn't the ES rest client being used as opposed to Apache HTTP? If so, I think these defaults line up with what is [documented](https://www.elastic.co/guide/en/elasticsearch/client/java-rest/current/_timeouts.html).

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestClientSetup.java (hunk 3 above)
Comment (twilmes) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestClientSetup.java]: Yes, good catch, bad copy-paste...

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java (hunk 1 above)
Comment (li-boxuan) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java:273]: Thanks, you are right. ES rest-client internally uses Apache HTTP client (we are using 4.x HTTP client) which [uses -1 as the default value ](https://github.com/apache/httpcomponents-client/blob/4.5.x/httpclient/src/main/java/org/apache/http/client/config/RequestConfig.java#L280-L292)(which is interpreted as an infinite timeout), but [ES overrides these values](https://github.com/elastic/elasticsearch/blob/master/client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java#L50-L51). So the default values indeed are 1000, but shall we use `RestClientBuilder.DEFAULT_CONNECT_TIMEOUT_MILLIS`?

Comment (twilmes) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java]: Yes, agreed, changed to the constant values.

**Review by li-boxuan — APPROVED** [review]
LGTM 👍

**Pull request merged**